### PR TITLE
add enableScala option

### PR DESCRIPTION
**Pull request by hturki**
hey @uschi2000 this is what I was talking about re: the scala support a week or two back

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/http-remoting/418)

### CODE REVIEW

**Comment by markelliot** [comment]
low confidence we want to support this; do you really *have* to use Scala objects in your API?

**Comment by hturki** [comment]
per discussion will move to support to a separate jaxrs-scala-clients project

File: jaxrs-clients/src/main/java/com/palantir/remoting2/jaxrs/AbstractFeignJaxRsClientBuilder.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+import feign.okhttp.OkHttpClient;
+import feign.slf4j.Slf4jLogger;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import okhttp3.ConnectionPool;
+import okhttp3.ConnectionSpec;
+import okhttp3.Credentials;
+import okhttp3.TlsVersion;
+
+public abstract class AbstractFeignJaxRsClientBuilder extends ClientBuilder {
```
Comment (uschi2000) [inline jaxrs-clients/src/main/java/com/palantir/remoting2/jaxrs/AbstractFeignJaxRsClientBuilder.java]: would prefer to not make this public

Comment (uschi2000) [reply jaxrs-clients/src/main/java/com/palantir/remoting2/jaxrs/AbstractFeignJaxRsClientBuilder.java]: (I don't want people to make their own subclasses outside this project.)

File: jaxrs-scala-clients/build.gradle [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+dependencies {
+    compile project(":jaxrs-clients")
+    compile "com.fasterxml.jackson.module:jackson-module-scala_2.11"
+
+    // Keep in sync with the dependencies excluded from shadowJar.
+    shadow project(":jaxrs-clients")
+    shadow "com.fasterxml.jackson.module:jackson-module-scala_2.11"
+}
+
+shadowJar {
```
Comment (uschi2000) [inline jaxrs-scala-clients/build.gradle]: :( this is pretty unmaintainable already, but duplicating this logic into two places makes it worse. can we at least factor out?

**Review by uschi2000 — COMMENTED** [review]
I'm not convinced this is the right decomp. Maybe it's easier to add a flag to JaxRsClientBuilder? let's discuss in person, ja?

**Comment by hturki** [comment]
@uschi2000 I actually initially wrote this via the flag approach :) but @markelliot raised some concerns around making the default client fatter / adding a dependency on the scala library for example. but yeah let's just catch up next week in a higher bandwidth medium

File: jaxrs-scala-clients/build.gradle (hunk 2 above)
Comment (hturki) [reply jaxrs-scala-clients/build.gradle]: yeah, open to ideas on the cleanest way of going about this

File: jaxrs-clients/src/main/java/com/palantir/remoting2/jaxrs/AbstractFeignJaxRsClientBuilder.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+    private final ClientConfig config;
+
+    AbstractFeignJaxRsClientBuilder(ClientConfig config) {
+        this.config = config;
+        Preconditions.checkArgument(config.maxNumRetries() == 0,
+                "Connection-level retries are not supported by %s", JaxRsClient.class.getSimpleName());
+    }
+
+    protected abstract ObjectMapper getObjectMapper();
```
Comment (uschi2000) [inline jaxrs-clients/src/main/java/com/palantir/remoting2/jaxrs/AbstractFeignJaxRsClientBuilder.java:75]: this unfortunately makes it very very easy for anyone outside this library to fuck with the object mapper. there's probably no way around this :( maybe service loading for adding the scala support at runtime? at the minimum, please add a class-level comment saying, that nobody outside this library must implement this class. thoughts, @markelliot ?

**Comment by uschi2000** [comment]
shading logics looks OK.

File: jaxrs-clients/src/main/java/com/palantir/remoting2/jaxrs/AbstractFeignJaxRsClientBuilder.java (hunk 3 above)
Comment (uschi2000) [reply jaxrs-clients/src/main/java/com/palantir/remoting2/jaxrs/AbstractFeignJaxRsClientBuilder.java:75]: ping @markelliot

**Comment by uschi2000** [comment]
Closing in favour of https://github.com/palantir/http-remoting/pull/439

**Pull request closed without merging**